### ISSUE TITLE
tf.layers.batch_normalization deprecation is a regression

### ISSUE THREAD

**Issue by nwestlake** [tensorflow/tensorflow#25525]
**System information**
- TensorFlow version (use command below):
('v1.12.0-7360-g5076adf64a', '1.13.0-dev20190204')

**Describe the behaviour**
[tensorflow/tensorflow/python/layers/normalization.py](https://github.com/tensorflow/tensorflow/blob/r1.13/tensorflow/python/layers/base.py) has a deprecation warning: (Use keras.layers.batch_normalization instead.')

However, the BatchNormalization class in that file also inherits from base.Layer. In particular, this means it updates UPDATE_OPS which the keras layer doesn't. 

The workaround is cumbersome. One is use the following code:
    
    def batch_norm(net_input, is_training, trainable=True, momentum=0.99,
               name="BatchNorm"):
    layer = tf.keras.layers.BatchNormalization(
        epsilon=1E-5, center=True, scale=True, momentum=momentum,
        trainable=trainable, name=name)
    add_elements_to_collection(layer.updates, tf.GraphKeys.UPDATE_OPS)

    return layer.apply(net_input, training=is_training)

With a separate method

    def add_elements_to_collection(elements, collection_list):
        elements = tf.nest.flatten(elements)
        collection_list = tf.nest.flatten(collection_list)
        for name in collection_list:
           collection = tf.get_collection_ref(name)
           collection_set = set(collection)
           for element in elements:
               if element not in collection_set:
                    collection.append(element)

The other workaround is to create a separate list, to be used instead of tf.GraphKeys.UPDATE_OPS.

In my view, this is a regression bug as avoiding the use of the deprecated methods involves the loss of useful feature and requirement of extra code.  The canned model such as [ResNet](https://github.com/tensorflow/models/blob/master/official/resnet/resnet_model.py) will be affected by this.

**Comment by karmel** [comment]
In TensorFlow 2.0, there will be no collections, as these represent global state that we are moving away from. The keras layers operate in the way that 2.0 at large will operate-- objects like variables and ops should be tracked and passed around explicitly using their Python handles. 

There are a couple of ways to think about the above: 
* Simplify the workaround using tf.add_to_collection.
* Ignore the deprecation warning and stick with the old batch_norm-- this will work for the life of TF 1.x
* Update to the Keras layers and don't rely on collections to run update ops, as this will help prepare you for the future with 2.0. It's hard to advise on how to do this given the code snippet above, but with more context about the systems/APIs/needs of what your doing, I could give more pointers as to how such code would look in the new world.

Let me know if the above makes sense.

**Comment by nwestlake** [comment]
So how does this work with estimator and model_fn? Will users have to move away from having a model_fn as an actual function which calls functions and have to create a class whose only purpose is to additionally keep track of update_ops?

**Comment by karmel** [comment]
Keras APIs handle layer update ops as part of their fit and evaluate loops. However if you use the Estimator API with a Keras model as part of your model function or any low level TensorFlow APIs, you will need to add these layer updates explicitly in your training loop to the train_op. For example, consider this example which uses the Estimator API:

```
def model_fn(features, labels):
  inputs = tf.keras.layers.Input(shape=...)
  output = tf.keras.layers.BatchNormalization(..., name=...)(inputs, training=....)
 
  model = tf.keras.models.Model(inputs, output, name=....)
  logits = model(features)
  ...
  
  # Find the layer in the model corresponding to the batchnorm layer and add the    
  # layer updates explicitly.
  update_ops = batch_norm_layer.get_updates_for(features) 
  
  minimize_op =  optimizer.apply_gradients(...)
  train_op = tf.group(minimize_op, update_ops)
 
  return EstimatorSpec(... train_op=train_op)
```

The new Estimator Head that is being worked on will also take update_ops as an explicit kwarg.

Note that if you use model_to_estimator to convert your existing Keras model to an Estimator then layer updates are handled for you by the API, and you do not need to add anything extra.

Does that make sense?

**Comment by karmel** [comment]
CC @dynamicwebpaige -- this is a FAQ; is there a place we are compiling FAQs?

**Comment by nwestlake** [comment]
Thanks for the reply.

I think my use case is the use of a model_fn without Keras rather than using a Keras model in an estimator. Currently, users are being told that the use of batch_normalization is deprecated and being directed to the Keras layer. The problems are two fold:

1. (communication "gotcha") The alternative does not change UPDATE_OPS but there is no warning in the deprecation message.
2. The alternative is not a drop in replacement as some users (like me), will have avoided using class and simply calling nested functions such that the model_fn is used as a pure function. (I understand it isn't due to the global state.) 

So I can think of two solutions.
1. Pass in a list to add update_ops, into every method which might have batch normalized in the function nest. (Use a decorator?)
2. Use something akin to tf.variable_scope and tf.device. However, are these being deprecated as part of the same changes?

What is the recommended solution for those not using Keras and who are using functions rather than trying to build the model in a class?

**Comment by karmel** [comment]
Re: (1) I agree that additional notes in the error would be helpful-- would you be interested in sending a PR that reflects your preferred messaging?

With regard to making sure these are called, one option is to use the @tf.function decorator. This decorator introduces implied control dependencies, so update ops inside should run as expected. There are restrictions on the decorator-- ie, variables can only be created during the first pass of the function, and otherwise must be passed in as parameters-- but you can read more and consider whether this is useful for you here: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/eager/def_function.py#L703

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by tensorflowbutler** [comment]
It has been 29 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by jvishnuvardhan** [comment]
Closing due to lack of recent activity. Please update the issue when new information becomes available, and we will open a new issue. Thanks!

**Comment by nwestlake** [comment]
Apologies for the delay. I am still puzzled with the use of Keras, having previously relied on estimator. 

Please can I refer to [the official Keras resnet model](https://github.com/tensorflow/models/blob/master/official/resnet/keras/resnet_model.py):

    x = layers.BatchNormalization(axis=bn_axis,
                                  momentum=BATCH_NORM_DECAY,
                                  epsilon=BATCH_NORM_EPSILON,
                                  name=bn_name_base + '2a')(x)
    x = layers.Activation('relu')(x)

 Is this a mistake because the updates from the BatchNormalisation layer are not updated? Is the training script in effect cheating and relying on training-like BatchNormlisation for evaluation?

**Comment by isaprykin** [comment]
If `Model.fit()` or `Model.evaluate()` API is used then the updates are being taken care of.

**Comment by mohapatras** [comment]
> Keras APIs handle layer update ops as part of their fit and evaluate loops. However if you use the Estimator API with a Keras model as part of your model function or any low level TensorFlow APIs, you will need to add these layer updates explicitly in your training loop to the train_op. For example, consider this example which uses the Estimator API:
> 
> ```
> def model_fn(features, labels):
>   inputs = tf.keras.layers.Input(shape=...)
>   output = tf.keras.layers.BatchNormalization(..., name=...)(inputs, training=....)
>  
>   model = tf.keras.models.Model(inputs, output, name=....)
>   logits = model(features)
>   ...
>   
>   # Find the layer in the model corresponding to the batchnorm layer and add the    
>   # layer updates explicitly.
>   update_ops = batch_norm_layer.get_updates_for(features) 
>   
>   minimize_op =  optimizer.apply_gradients(...)
>   train_op = tf.group(minimize_op, update_ops)
>  
>   return EstimatorSpec(... train_op=train_op)
> ```
> 
> The new Estimator Head that is being worked on will also take update_ops as an explicit kwarg.
> 
> Note that if you use model_to_estimator to convert your existing Keras model to an Estimator then layer updates are handled for you by the API, and you do not need to add anything extra.
> 
> Does that make sense?


@karmel On the part where using model_to_estimator handles the update ops for the user. Just to clarify, while training a user will directly runs train_and_evaluate() and the "update_ops" get handled automatically ?

**Comment by iperov** [comment]
my solution to make BatchNormalization works properly , i.e. updating moving_mean and moving_variance, if you are using K.function API

```
def xor_list(lst1, lst2): 
    return  [value for value in lst1+lst2 if (value not in lst1) or (value not in lst2)  ]  

#decoder - is keras model
decoder_def_updates = decoder.updates.copy()
loss = decoder (code)

#fetching only new appeared updates 
decoder_updates = xor_list (decoder_def_updates, self.decoder.updates )

#now we gathered all updates of inner layers in decoder model
...
#pass these updates together with optimizer updates
K.function ([...inputs...],[loss], optimizer.get_updates(loss, weights) + decoder_updates )
```
